File: integrations/aws-strands/typescript/examples/server/model-factory.ts
/**
 * Shared model factory for Strands TypeScript examples.
 *
 * Mirrors `python/examples/server/model_factory.py` field-for-field: same
 * `MODEL_PROVIDER` env-var shape, same defaults, same model IDs. The CI dojo
 * runner injects `OPENAI_BASE_URL=http://localhost:5555/v1` + a mock API key,
 * so the default `openai` provider routes to the aimock server automatically.
 *
 * Reasoning summaries are opt-in via `{ reasoning: true }`. Mirrors the
 * mastra/langgraph TS dojos, where only the `agentic-chat-reasoning` demo
 * configures a reasoning model. Leaving reasoning on by default produces
 * `reasoningBlock` content in assistant turns that the OpenAI Responses API
 * cannot replay across multi-turn conversations, which breaks any demo that
 * triggers a tool-use loop (e.g. `tool-based-generative-ui`).
 *
 * Supported providers: `openai` (default), `anthropic`, `gemini`, `bedrock`.
 */

import type { Model } from "@strands-agents/sdk";

export interface CreateModelOptions {
  /**
   * Request reasoning/thinking content from the provider. Defaults to `false`.
   * Only enable for demos that explicitly render reasoning in the UI — the
   * Responses API drops reasoning blocks across multi-turn conversations.
   */
  reasoning?: boolean;
  /**
   * OpenAI API mode. Defaults to the SDK default (Responses). Pass `"chat"`
   * for demos that need tool-call ARGUMENTS to stream incrementally — the
   * Strands Responses adapter buffers `function_call_arguments.delta` and only
   * emits the complete toolUse at `…arguments.done`, so e.g. A2UI progressive
   * surface painting never streams on the Responses API.
   */
  openaiApi?: "chat" | "responses";
}

export async function createModel(
  options: CreateModelOptions = {},
): Promise<Model> {
  const provider = (process.env.MODEL_PROVIDER ?? "openai").toLowerCase();
  const reasoning = options.reasoning ?? false;

  if (provider === "openai") {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
      throw new Error(
        "OPENAI_API_KEY environment variable is required when MODEL_PROVIDER=openai. " +
          "Set it in your .env file or environment.",
      );
    }
    const { OpenAIModel } = await import("@strands-agents/sdk/models/openai");
    // OPENAI_BASE_URL routes through aimock during e2e tests. The default
    // Responses API surfaces fixture `reasoning` content for the
    // `/agentic-chat-reasoning` demo.
    const baseURL = process.env.OPENAI_BASE_URL;
    return new OpenAIModel({
      apiKey,
      modelId: process.env.MODEL_ID ?? "gpt-5.4",
      ...(options.openaiApi ? { api: options.openaiApi } : {}),
      ...(reasoning
        ? { params: { reasoning: { effort: "medium", summary: "auto" } } }
        : {}),
      ...(baseURL ? { clientConfig: { baseURL } } : {}),
    });
  }

  if (provider === "anthropic") {
    const apiKey = process.env.ANTHROPIC_API_KEY;
    if (!apiKey) {
      throw new Error(
        "ANTHROPIC_API_KEY environment variable is required when MODEL_PROVIDER=anthropic. " +
          "Set it in your .env file or environment.",
      );
    }
    const { AnthropicModel } = await import(
      "@strands-agents/sdk/models/anthropic"
    );
    return new AnthropicModel({
      apiKey,
      modelId: process.env.MODEL_ID ?? "claude-sonnet-4-6",
    });
  }

  if (provider === "gemini") {
    const apiKey = process.env.GOOGLE_API_KEY;
    if (!apiKey) {
      throw new Error(
        "GOOGLE_API_KEY environment variable is required when MODEL_PROVIDER=gemini. " +
          "Set it in your .env file or environment.",
      );
    }
    const { GoogleModel } = await import("@strands-agents/sdk/models/google");
    return new GoogleModel({
      apiKey,
      modelId: process.env.MODEL_ID ?? "gemini-2.5-flash",
    });
  }

  if (provider === "bedrock") {
    const { BedrockModel } = await import("@strands-agents/sdk");
    // Anthropic-on-Bedrock surfaces reasoning via the
    // `additionalModelRequestFields.thinking` block. `temperature` must be 1
    // when thinking is enabled.
    // https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-anthropic-claude-messages-extended-thinking.html
    return new BedrockModel({
      modelId: process.env.MODEL_ID ?? "global.anthropic.claude-sonnet-4-6",
      ...(reasoning
        ? {
            temperature: 1,
            additionalRequestFields: {
              thinking: { type: "enabled", budget_tokens: 2000 },
            },
          }
        : {}),
    });
  }

  throw new Error(
    `Unknown MODEL_PROVIDER: ${provider}. Supported: openai, anthropic, gemini, bedrock`,
  );
}

File: integrations/aws-strands/python/examples/server/model_factory.py
"""Shared model factory for Strands examples.

Supports OpenAI, Anthropic, and Gemini via MODEL_PROVIDER env var.
Defaults to OpenAI.
"""
import os
import logging

logger = logging.getLogger(__name__)


def create_model(openai_api: str = "responses"):
    """Create a Strands model based on MODEL_PROVIDER env var.

    Supported providers: openai (default), anthropic, gemini

    ``openai_api`` selects the OpenAI API mode. The default Responses API
    surfaces reasoning summaries but buffers tool-call argument deltas until
    the call completes; pass ``"chat"`` for demos that need tool-call ARGUMENTS
    to stream incrementally (e.g. A2UI progressive surface painting).
    """
    provider = os.getenv("MODEL_PROVIDER", "openai").lower()

    if provider == "openai":
        api_key = os.getenv("OPENAI_API_KEY")
        if not api_key:
            raise ValueError(
                "OPENAI_API_KEY environment variable is required when MODEL_PROVIDER=openai. "
                "Set it in your .env file or environment."
            )
        if openai_api == "chat":
            from strands.models.openai import OpenAIModel
            return OpenAIModel(
                client_args={
                    "api_key": api_key,
                },
                model_id=os.getenv("MODEL_ID", "gpt-5.4"),
            )
        from strands.models.openai_responses import OpenAIResponsesModel
        return OpenAIResponsesModel(
            client_args={
                "api_key": api_key,
            },
            model_id=os.getenv("MODEL_ID", "gpt-5.4"),
            params={
                "reasoning": {"effort": "medium", "summary": "auto"},
            }
        )
    elif provider == "anthropic":
        api_key = os.getenv("ANTHROPIC_API_KEY")
        if not api_key:
            raise ValueError(
                "ANTHROPIC_API_KEY environment variable is required when MODEL_PROVIDER=anthropic. "
                "Set it in your .env file or environment."
            )
        from strands.models.anthropic import AnthropicModel
        return AnthropicModel(
            client_args={
                "api_key": api_key,
            },
            model_id=os.getenv("MODEL_ID", "claude-sonnet-4-6"),
            params={
                "budget_tokens": 5000,
            }
        )
    elif provider == "gemini":
        api_key = os.getenv("GOOGLE_API_KEY")
        if not api_key:
            raise ValueError(
                "GOOGLE_API_KEY environment variable is required when MODEL_PROVIDER=gemini. "
                "Set it in your .env file or environment."
            )
        from strands.models.gemini import GeminiModel
        return GeminiModel(
            client_args={
                "api_key": api_key,
            },
            model_id=os.getenv("MODEL_ID", "gemini-2.5-flash"),
            params={
                "temperature": 0.7,
                "max_output_tokens": 2048,
            }
        )
    else:
        raise ValueError(f"Unknown MODEL_PROVIDER: {provider}. Supported: openai, anthropic, gemini")
